stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -125,275 +125,185 @@ end
-    #=> [0, 1, tau1, tau0, tau_ptr+4, ...]
-
-    # 2) Compute tau^i
-    repeat.512
-        dupw ext2mul
-
-        movup.3 movup.3
-
-        dup.6 add.4 swap.7 mem_storew
-
-        drop drop
-    end
-
-    dropw
-    #=> [tau_ptr + 513*4, ...]
-end
-
-
-#! Sets the memory region `[ptr, ptr + 512*4)` to zero. The pointer c_ptr := ptr + 512*4 is returned
-#! to be used to store the hash-to-point polynomial of the message later on.
-#!
-#! Input: [ptr, ...]
-#! Output: [...]
-#!
-#! Cycles: ~3100
-export.set_to_zero
-    padw
-    repeat.512
-        dup.4 add.4 swap.5
-        mem_storew
-    end
-    dropw
-end
-
 #! Takes as input PK, the hash of the coefficients of the polynomial `h` representing the expanded
 #! public key, and a pointer to the memory location where the coefficients of the polynomial `h`
 #! will be stored.
 #! The procedure loads `h` from the advice stack and compares its hash with the provided hash `PK`.
 #! It then loads the polynomial `s2` representing the signature from the advice stack and lays it
 #! in memory right after `h`.
-#! It then loads the claimed polynomial `h * s2` in Z_Q[x] where Q is the Miden VM prime from
-#! the advice stack and lays it right after `s2`.
+#! After that, it loads the claimed polynomial `h * s2` in Z_Q[x] where Q is the Miden VM prime
+#! from the advice stack and lays it right after `s2`.
 #! The hash of `h`, `s2` and the claimed product is also computed and the first two field elements
-#! of the digest (i.e., the Fiat-Shamir challenge) are returned on the stack alongside
-#! the incremented pointer.
-#!
-#! Input: [ptr, PK, ...]
-#! Output: [tau1, tau0, ptr + 512*4 ...]
-#!
-#! Cycles: 5049
-export.load_h_s2_and_product.4
+#! of the digest (i.e., the Fiat-Shamir challenge) are used in order to check that 
+#! pi == h * s2 in Z_Q[x] by evaluating both sides at the random point. Note that since we need
+#! to hash the polynomials before computing their evaluations, the procedure receives
+#! the evaluation point non-deterministically so that it can perform the evaluations while hashing
+#! is still going on. After hashing of the polynomials is finished, the procedure checks that
+#! the evaluation point was indeed derived correctly from the hash of all 3 polynomials.
+#!
+#! Inputs:
+#!   Operand stack: [ptr, PK, ...]
+#!   Advice stack:  [tau0, tau1, h_0, ..., h_511, s2_0, ..., s2_511, pi_0, ..., pi_1022, ...]
+#! Outputs:
+#!   Operand stack: []
+#!   Advice stack:  []
+#!
+#! Cycles: 4430
+export.load_h_s2_and_product.12
+    # We will store in local addresses:
+    # 1. 0..4 the public key,
+    # 2. 4..8 the inverse of the evaluation point tau as [tau_inv0, tau_inv1, 0, 0],
+    # 3. 8..12 the evaluation point tau as [tau0, tau1, 0, 0],
+    
+    # 1) Set up the stack for loading the coefficients of the polynomials, evaluating and hashing them
+    
+    ## a) Set up the accumulator for `horner_eval_base` and the memory pointers
+    push.0.0
+    locaddr.4
+    movup.3
```

Comment:
Could we add stack state after this instruction?

Reply:
done